src/java.base/share/classes/sun/nio/fs/AbstractWatchKey.java

```diff
@@ -28,16 +28,39 @@
 import java.nio.file.*;
 import java.util.*;
 
+import jdk.internal.util.ArraysSupport;
+import sun.security.action.GetPropertyAction;
+
 /**
  * Base implementation class for watch keys.
  */
 
 abstract class AbstractWatchKey implements WatchKey {
 
+    private static final int DEFAULT_MAX_EVENT_LIST_SIZE = 512;
+
     /**
-     * Maximum size of event list (in the future this may be tunable)
+     * Maximum size of event list before dropping events and signalling OVERFLOW
      */
-    static final int MAX_EVENT_LIST_SIZE    = 512;
+    @SuppressWarnings("removal")
+    static final int MAX_EVENT_LIST_SIZE;
+    static {
+        String rawValue = GetPropertyAction.privilegedGetProperty(
+            "jdk.nio.file.WatchService.maxEventsPerPoll",
+            String.valueOf(DEFAULT_MAX_EVENT_LIST_SIZE));
+        int intValue;
+        try {
+            // Clamp to max array length to signal OVERFLOW and drop events
+            // before OOMing.
+            intValue = Math.clamp(
+                Long.decode(rawValue),
+                1,
+                ArraysSupport.SOFT_MAX_ARRAY_LENGTH);
```

Comment:
The use of ArraysSupport.SOFT_MAX_ARRAY_LENGTH to limit the value seems a bit overkill here. It's okay, just overkill.

Reply:
I am also okay with it. This might be a little simpler, however:
```
    static final int MAX_EVENT_LIST_SIZE;
    static {
        int intValue = DEFAULT_MAX_EVENT_LIST_SIZE;
        String rawValue = GetPropertyAction.privilegedGetProperty
            ("jdk.nio.file.WatchService.maxEventsPerPoll");
        if (rawValue != null) {
            try {
                intValue = Integer.decode(rawValue);
                if (intValue <= 0)
                    intValue = 1;
            } catch (NumberFormatException ignored) {
            }
        }
        MAX_EVENT_LIST_SIZE = intValue;
    }
```
Given that a lot of tests have already been run, it might be better just to go with 0703b3f499ee32a7efd780745c42b42131dfef85. I will plan to approve that tomorrow unless there is an objection.

As the CSR was approved, once the review is done, this can be integrated and then we will sponsor it.